### DISCUSSION TITLE
Temporary non-QWERTY non-Latin keyboard layout support

### DISCUSSION THREAD

**Discussion by eugenesvk** [helix-editor/helix#7097] [Show and tell]
There is an issue https://github.com/helix-editor/helix/issues/133 opened re. proper support of international keyboard layout support

Meanwhile I've ported my little Python scripts that allow you to add your custom keymap (map every qwerty key to some other key) and then use it to generate a new config file that includes existing config replaced/extended with the keys converted to this custom keymap.
  - min script https://github.com/eugenesvk/kbdHelix/blob/modifew/src/modify_config.toml.tmpl
  - conversion module where you'd actually add your alternative keymap https://github.com/eugenesvk/kbdHelix/blob/modifew/src/layout_convert.py
  - a `tomlkit` library for parsing TOML configs (not included)

While the script is part of a more complicated chezmoi-based build setup that allows you to do some sane config management for extensive remapping , you could also tweak [this line](https://github.com/eugenesvk/kbdHelix/blob/e144c830603322ab89fc60e288d7de71cb9d5680/src/modify_config.toml.tmpl#L29) to just read from an existing Helix config file and then redirect [this print](https://github.com/eugenesvk/kbdHelix/blob/e144c830603322ab89fc60e288d7de71cb9d5680/src/modify_config.toml.tmpl#L123) to create another file with which to replace your config (don't forget to back the original up) to ignore that complexity

P.S. This approach has two downsides
- For non-latin layouts keys that are not translated (e.g., space) that are used as submodes/menus get polluted with both Latin and non-Latin variants regardless of your layout
- You can't have the same symbol act differently depending on your active layout as ultimately the config is ignorant about layouts and only understands non-Latin input due to non-Latin characters the app receives

**Comment by VKondakoff** [comment]
@eugenesvk,

please, can you give a little bit more detail of how to use it? I installed `tomlkit`, downloaded both files to kbdHelix folder and tried to run `layout_convert.py`, expecting to see the output, where Cyrillic keys (`йцукен`) are mapped to Latin (`qwerty`). But nothing happens. Should I edit the scripts first somehow? 

**Edit:** Ah, got it. I should run `modify_config.toml.tmpl`, and not the `layout_convert.py`. But running `modify_config.toml.tmpl` gives me a bunch of errors:
<img width="570" alt="Снимок экрана 2024-04-03 в 12 54 16" src="https://github.com/helix-editor/helix/assets/230824/856cd8bf-17f1-46ec-9c63-b7c440c824a3">

Thank you!

PS: shouldn't [this line](https://github.com/eugenesvk/kbdHelix/blob/e144c830603322ab89fc60e288d7de71cb9d5680/src/modify_config.toml.tmpl#L12) be commented?

**Reply by eugenesvk** [reply]
It means your config has no `[keys.normal]`, so the parser can't read those modes from it, did you read the full config to the string (the script doesn't do that currently, it's fed data by chezmoi)?

Replace `cfg_def_s` with something like this to read from your config file and try again
```py
cfg_def_f = r"""path/to/config.toml"""
cfg_def_s = ""
with open(cfg_def_f,'r') as file:
  cfg_def_s = file.read()
```


fixed the comment (template was ignoring it inside its own comment)

**Comment by VKondakoff** [comment]
Aha, got it. Thank you, it works now: if there are some `qwerty` keymaps in `config.toml`, they are mapped to their `йцукен` equivalent. But this is not very useful, because default Helix keymaps are not in the config, so they are not mapped. Like these (no one adds default keymaps to the config): 
```
"k" = "move_line_up" 
"h" = "move_char_right"
"j" = "move_line_down" 
```
I was hoping to find a way to remap _every_ existing Helix Normal mode keymap to its `йцукен` equivalent... Any idea if this is possible? I really don't want to move every Normal mode keymap from [the docs](https://docs.helix-editor.com/master/keymap.html) to the config...

**Reply by eugenesvk** [reply]
> no one adds default keymaps to the config)

I did :) since all of these defaults are bad

there is some default config https://github.com/eugenesvk/kbdHelix/blob/main/helper/config_default.toml form an unknow version, and maybe you could find a better version here https://github.com/helix-editor/helix/issues/1210

Or you could move from the source code https://github.com/helix-editor/helix/blob/master/helix-term/src/keymap/default.rs or maybe even write a simple parsing script to generate toml :)

**Reply by VKondakoff** [reply]
Hey, that's great! Thank you for help!
